File: 2018/02/main.py
from itertools import groupby

def different_char_indexes(s1, s2):
    indexes = []
    for index, char in enumerate(s1):
        if s2[index] != char:
            indexes.append(index)
    return indexes

with open('./input.txt') as f:
    ids = list(f)
    two = 0
    three = 0

    for id in ids:
        counts = [id.count(char) for char in id]
        if 2 in counts:
            two += 1
        if 3 in counts:
            three += 1

    print("Part 1:")
    print(two * three)

    result = None
    for id in ids:
        for other in ids:
            diffs = different_char_indexes(id, other)
            if len(diffs) == 1:
                index = diffs[0]
                result = id[:index] + id[(index + 1):]
                break

        if result:
            break

    print("Part 2:")
    print(result)

File: 2018/rust/02/main.rs
use std::fs::File;
use std::io::prelude::*;

fn different_char_count(s1: &str, s2: &str) -> i32 {
    s1.chars()
        .zip(s2.chars())
        .fold(0, |acc, (x, y)| if x == y { acc } else { acc + 1 })
}

fn main() {
    let mut f = File::open("input.txt").expect("input not found");
    let mut input = String::new();
    f.read_to_string(&mut input).expect("coudln't read input");
    input.pop(); // remove trailing newline

    let ids: Vec<&str> = input.split_whitespace().collect();

    let mut double_occurances = 0;
    let mut triple_occurances = 0;
    let mut found_pair = false;
    let mut pair1 = "";
    let mut pair2 = "";

    for id in ids.iter() {
        let mut chars: Vec<char> = id.chars().collect();
        chars.sort();
        chars.dedup();

        let char_counts: Vec<usize> = chars.iter()
            .map(|c| id.matches(c.to_string().as_str()).count())
            .collect();

        if char_counts.contains(&2) { double_occurances += 1 }
        if char_counts.contains(&3) { triple_occurances += 1 }

        if found_pair { continue }
        ids.iter()
            .find(|other| different_char_count(id, other) == 1)
            .map(|other| {
                pair1 = id;
                pair2 = other;
                found_pair = true;
            });
    }

    println!("Part 1: {}", double_occurances * triple_occurances);

    let chars1: Vec<char> = pair1.chars().collect();
    let chars2: Vec<char> = pair2.chars().collect();
    let common: String = chars1.iter().filter(|c| chars2.contains(&c)).collect();
    println!("{}", common);
}
